File: flutter_app/lib/services/recurring.dart
import '../models/finance.dart';

/// Pure helpers mirroring the React `lib/finance/recurring.ts` logic so both
/// apps surface the same recurring occurrences. All date math is done on plain
/// YYYY-MM-DD strings in UTC to avoid timezone drift.

String _pad(int n) => n.toString().padLeft(2, '0');

String _toIso(DateTime d) =>
    '${d.year.toString().padLeft(4, '0')}-${_pad(d.month)}-${_pad(d.day)}';

DateTime _parseIso(String iso) {
  final parts = iso.split('-').map(int.parse).toList();
  return DateTime.utc(parts[0], parts.length > 1 ? parts[1] : 1,
      parts.length > 2 ? parts[2] : 1);
}

int _daysInMonth(int year, int month1) => DateTime.utc(year, month1 + 1, 0).day;

class RecurringOccurrence {
  RecurringOccurrence(this.dateISO, this.periodKey);
  final String dateISO;
  final String periodKey;
}

/// Most recent occurrence on or before [todayISO] for a rule, or null when the
/// rule has no occurrence at/before today (e.g. a biweekly anchor in future).
RecurringOccurrence? lastDueOccurrence(RegularPayment rule, String todayISO) {
  final today = _parseIso(todayISO);

  switch (rule.frequency) {
    case RecurringFrequency.monthly:
      final y = today.year;
      final m = today.month;
      final day = rule.dueDate.clamp(1, _daysInMonth(y, m));
      var occ = DateTime.utc(y, m, day);
      if (occ.isAfter(today)) {
        final py = m == 1 ? y - 1 : y;
        final pm = m == 1 ? 12 : m - 1;
        final pday = rule.dueDate.clamp(1, _daysInMonth(py, pm));
        occ = DateTime.utc(py, pm, pday);
      }
      return RecurringOccurrence(
          _toIso(occ), '${occ.year.toString().padLeft(4, '0')}-${_pad(occ.month)}');

    case RecurringFrequency.yearly:
      final month1 = (rule.month ?? 1).clamp(1, 12);
      final y = today.year;
      final day = rule.dueDate.clamp(1, _daysInMonth(y, month1));
      var occ = DateTime.utc(y, month1, day);
      if (occ.isAfter(today)) {
        final py = y - 1;
        final pday = rule.dueDate.clamp(1, _daysInMonth(py, month1));
        occ = DateTime.utc(py, month1, pday);
      }
      return RecurringOccurrence(
          _toIso(occ), occ.year.toString().padLeft(4, '0'));

    case RecurringFrequency.weekly:
      // Dart: Mon=1..Sun=7. Convert to JS-style Sun=0..Sat=6.
      final todayDow = today.weekday % 7;
      final target = ((rule.weekday ?? todayDow) % 7 + 7) % 7;
      final diff = (todayDow - target + 7) % 7;
      final occ = today.subtract(Duration(days: diff));
      return RecurringOccurrence(_toIso(occ), _toIso(occ));

    case RecurringFrequency.biweekly:
      final anchorISO = rule.anchorDate;
      if (anchorISO == null) return null;
      final anchor = _parseIso(anchorISO);
      if (anchor.isAfter(today)) return null;
      final periods = today.difference(anchor).inDays ~/ 14;
      final occ = anchor.add(Duration(days: periods * 14));
      return RecurringOccurrence(_toIso(occ), _toIso(occ));
  }
}

String refFor(String ruleId, String periodKey) => '$ruleId:$periodKey';

bool isOccurrencePosted(
    List<Transaction> transactions, String ruleId, String periodKey) {
  final ref = refFor(ruleId, periodKey);
  return transactions.any((t) => t.recurringRef == ref);
}

class DueRecurring {
  DueRecurring(this.rule, this.periodKey, this.dateISO, this.ref);
  final RegularPayment rule;
  final String periodKey;
  final String dateISO;
  final String ref;
}

/// Active rules whose latest occurrence is due, not yet posted, and not skipped.
List<DueRecurring> getDueRecurring(
  List<RegularPayment> rules,
  List<Transaction> transactions,
  List<RecurringSkip> skips,
  String todayISO,
) {
  final skipSet = skips.map((s) => refFor(s.ruleId, s.periodKey)).toSet();
  final out = <DueRecurring>[];
  for (final rule in rules) {
    if (!rule.isActive) continue;
    final occ = lastDueOccurrence(rule, todayISO);
    if (occ == null) continue;
    final ref = refFor(rule.id, occ.periodKey);
    if (skipSet.contains(ref)) continue;
    if (isOccurrencePosted(transactions, rule.id, occ.periodKey)) continue;
    out.add(DueRecurring(rule, occ.periodKey, occ.dateISO, ref));
  }
  out.sort((a, b) => a.dateISO.compareTo(b.dateISO));
  return out;
}

File: src/lib/finance/recurring.ts
import type { RegularPayment, RecurringFrequency, RecurringSkip, Transaction } from '../../types';

// All date math is done on plain YYYY-MM-DD strings in UTC to avoid timezone drift.

function pad(n: number): string {
  return n.toString().padStart(2, '0');
}

export function toISO(date: Date): string {
  return `${date.getUTCFullYear()}-${pad(date.getUTCMonth() + 1)}-${pad(date.getUTCDate())}`;
}

function parseISO(iso: string): Date {
  const [y, m, d] = iso.split('-').map(Number);
  return new Date(Date.UTC(y, (m || 1) - 1, d || 1));
}

function daysInMonth(year: number, month1: number): number {
  // month1 is 1-12
  return new Date(Date.UTC(year, month1, 0)).getUTCDate();
}

export function frequencyOf(rule: RegularPayment): RecurringFrequency {
  return rule.frequency || 'monthly';
}

/**
 * Most recent occurrence on or before `today` for a recurring rule.
 * Returns null if the rule has no occurrence at/before today (e.g. anchor in future).
 */
export function lastDueOccurrence(
  rule: RegularPayment,
  todayISO: string,
): { dateISO: string; periodKey: string } | null {
  const today = parseISO(todayISO);
  const freq = frequencyOf(rule);

  if (freq === 'monthly') {
    const y = today.getUTCFullYear();
    const m = today.getUTCMonth() + 1; // 1-12
    const day = Math.min(Math.max(rule.dueDate || 1, 1), daysInMonth(y, m));
    let occ = new Date(Date.UTC(y, m - 1, day));
    if (occ.getTime() > today.getTime()) {
      // This month's occurrence is still in the future → use previous month.
      const py = m === 1 ? y - 1 : y;
      const pm = m === 1 ? 12 : m - 1;
      const pday = Math.min(Math.max(rule.dueDate || 1, 1), daysInMonth(py, pm));
      occ = new Date(Date.UTC(py, pm - 1, pday));
    }
    const dateISO = toISO(occ);
    return { dateISO, periodKey: `${occ.getUTCFullYear()}-${pad(occ.getUTCMonth() + 1)}` };
  }

  if (freq === 'yearly') {
    const month1 = Math.min(Math.max(rule.month || 1, 1), 12);
    const y = today.getUTCFullYear();
    const day = Math.min(Math.max(rule.dueDate || 1, 1), daysInMonth(y, month1));
    let occ = new Date(Date.UTC(y, month1 - 1, day));
    if (occ.getTime() > today.getTime()) {
      const py = y - 1;
      const pday = Math.min(Math.max(rule.dueDate || 1, 1), daysInMonth(py, month1));
      occ = new Date(Date.UTC(py, month1 - 1, pday));
    }
    return { dateISO: toISO(occ), periodKey: `${occ.getUTCFullYear()}` };
  }

  if (freq === 'weekly') {
    const target = ((rule.weekday ?? today.getUTCDay()) % 7 + 7) % 7;
    const diff = (today.getUTCDay() - target + 7) % 7; // days since last target weekday
    const occ = new Date(today.getTime() - diff * 86400000);
    return { dateISO: toISO(occ), periodKey: toISO(occ) };
  }

  // biweekly: every 14 days anchored on anchorDate
  const anchorISO = rule.anchorDate;
  if (!anchorISO) return null;
  const anchor = parseISO(anchorISO);
  if (anchor.getTime() > today.getTime()) return null;
  const periods = Math.floor((today.getTime() - anchor.getTime()) / (14 * 86400000));
  const occ = new Date(anchor.getTime() + periods * 14 * 86400000);
  return { dateISO: toISO(occ), periodKey: toISO(occ) };
}

/**
 * Earliest occurrence on or after `today` for a recurring rule — the next
 * time the rule will come due. Returns null only when the cadence cannot be
 * resolved (it never does for the supported frequencies).
 */
export function nextDueOccurrence(
  rule: RegularPayment,
  todayISO: string,
): { dateISO: string; periodKey: string } | null {
  const today = parseISO(todayISO);
  const freq = frequencyOf(rule);

  if (freq === 'monthly') {
    const y = today.getUTCFullYear();
    const m = today.getUTCMonth() + 1;
    const day = Math.min(Math.max(rule.dueDate || 1, 1), daysInMonth(y, m));
    let occ = new Date(Date.UTC(y, m - 1, day));
    if (occ.getTime() < today.getTime()) {
      const ny = m === 12 ? y + 1 : y;
      const nm = m === 12 ? 1 : m + 1;
      const nday = Math.min(Math.max(rule.dueDate || 1, 1), daysInMonth(ny, nm));
      occ = new Date(Date.UTC(ny, nm - 1, nday));
    }
    return { dateISO: toISO(occ), periodKey: `${occ.getUTCFullYear()}-${pad(occ.getUTCMonth() + 1)}` };
  }

  if (freq === 'yearly') {
    const month1 = Math.min(Math.max(rule.month || 1, 1), 12);
    const y = today.getUTCFullYear();
    const day = Math.min(Math.max(rule.dueDate || 1, 1), daysInMonth(y, month1));
    let occ = new Date(Date.UTC(y, month1 - 1, day));
    if (occ.getTime() < today.getTime()) {
      const ny = y + 1;
      const nday = Math.min(Math.max(rule.dueDate || 1, 1), daysInMonth(ny, month1));
      occ = new Date(Date.UTC(ny, month1 - 1, nday));
    }
    return { dateISO: toISO(occ), periodKey: `${occ.getUTCFullYear()}` };
  }

  if (freq === 'weekly') {
    const target = ((rule.weekday ?? today.getUTCDay()) % 7 + 7) % 7;
    const diff = (target - today.getUTCDay() + 7) % 7; // days until next target weekday
    const occ = new Date(today.getTime() + diff * 86400000);
    return { dateISO: toISO(occ), periodKey: toISO(occ) };
  }

  // biweekly: every 14 days anchored on anchorDate
  const anchorISO = rule.anchorDate;
  if (!anchorISO) return null;
  const anchor = parseISO(anchorISO);
  if (anchor.getTime() >= today.getTime()) {
    return { dateISO: toISO(anchor), periodKey: toISO(anchor) };
  }
  const periods = Math.ceil((today.getTime() - anchor.getTime()) / (14 * 86400000));
  const occ = new Date(anchor.getTime() + periods * 14 * 86400000);
  return { dateISO: toISO(occ), periodKey: toISO(occ) };
}

export function refFor(ruleId: string, periodKey: string): string {
  return `${ruleId}:${periodKey}`;
}

export function isOccurrencePosted(
  transactions: Transaction[],
  ruleId: string,
  periodKey: string,
): boolean {
  const ref = refFor(ruleId, periodKey);
  return transactions.some((t) => t.recurringRef === ref);
}

export type DueRecurring = {
  rule: RegularPayment;
  periodKey: string;
  dateISO: string;
  ref: string;
};

/**
 * Active recurring rules whose latest occurrence is due (on/before today),
 * not yet posted, and not skipped — i.e. awaiting the user's confirmation.
 */
export function getDueRecurring(
  rules: RegularPayment[],
  transactions: Transaction[],
  skips: RecurringSkip[],
  todayISO: string,
): DueRecurring[] {
  const skipSet = new Set(skips.map((s) => refFor(s.ruleId, s.periodKey)));
  const out: DueRecurring[] = [];
  for (const rule of rules) {
    if (!rule.isActive) continue;
    const occ = lastDueOccurrence(rule, todayISO);
    if (!occ) continue;
    const ref = refFor(rule.id, occ.periodKey);
    if (skipSet.has(ref)) continue;
    if (isOccurrencePosted(transactions, rule.id, occ.periodKey)) continue;
    out.push({ rule, periodKey: occ.periodKey, dateISO: occ.dateISO, ref });
  }
  // Earliest due first.
  out.sort((a, b) => a.dateISO.localeCompare(b.dateISO));
  return out;
}
